File: simulator.py
import time
import random
import json
import urllib.request
import urllib.error
import threading
import sys
import msvcrt
from datetime import datetime

# Server details
API_URL = "https://smart-agri-platform-1.onrender.com/api/sensor-data" if "--cloud" in sys.argv else "http://localhost:10000/api/sensor-data"

# Simulation state
# Options: 'normal', 'heat', 'humidity', 'hypoxia', 'bioelectrical'
current_state = 'normal'

# Lock for sharing state between input thread and main loop
state_lock = threading.Lock()

def print_menu():
    print("=" * 60)
    print("  Agri AI IoT Hardware Simulator  ")
    print("=" * 60)
    print("Interactive Stress Controls:")
    print("  [n] - Reset to NORMAL state")
    print("  [t] - Inject HEAT stress (High Temperature)")
    print("  [h] - Inject DRY stress (Low Humidity)")
    print("  [o] - Inject HYPOXIA stress (Low Dissolved Oxygen)")
    print("  [s] - Inject BIOELECTRICAL stress (High Stress status)")
    print("  [q] - Quit Simulator")
    print("-" * 60)
    print("Press a key at any time to inject/change stress conditions.")
    print("=" * 60)

def read_keyboard():
    global current_state
    while True:
        if msvcrt.kbhit():
            ch = msvcrt.getch().decode('utf-8', errors='ignore').lower()
            if ch == 'q':
                print("\n[SIMULATOR] Exiting...")
                sys.exit(0)
            
            with state_lock:
                if ch == 'n':
                    current_state = 'normal'
                    print(f"\n[SIMULATOR] State changed to: NORMAL")
                elif ch == 't':
                    current_state = 'heat'
                    print(f"\n[SIMULATOR] State changed to: HEAT STRESS")
                elif ch == 'h':
                    current_state = 'humidity'
                    print(f"\n[SIMULATOR] State changed to: DRY STRESS (LOW HUMIDITY)")
                elif ch == 'o':
                    current_state = 'hypoxia'
                    print(f"\n[SIMULATOR] State changed to: HYPOXIA STRESS (LOW DO)")
                elif ch == 's':
                    current_state = 'bioelectrical'
                    print(f"\n[SIMULATOR] State changed to: BIOELECTRICAL STRESS")
        time.sleep(0.1)

def generate_telemetry(state):
    timestamp = datetime.utcnow().isoformat() + "Z"
    
    # Base normal levels
    temp = round(random.uniform(22.0, 26.0), 1)
    humidity = round(random.uniform(60.0, 75.0), 1)
    n = random.randint(110, 130)
    p = random.randint(25, 35)
    k = random.randint(85, 105)
    do_mg = round(random.uniform(6.5, 7.5), 2)
    voltage = random.randint(45, 55)
    plant_status = "Healthy"
    
    if state == 'heat':
        temp = round(random.uniform(32.0, 36.0), 1)
        voltage = random.randint(30, 42)
        plant_status = "Moderate Stress"
    elif state == 'humidity':
        humidity = round(random.uniform(40.0, 48.0), 1)
        voltage = random.randint(35, 45)
        plant_status = "Moderate Stress"
    elif state == 'hypoxia':
        do_mg = round(random.uniform(3.5, 4.5), 2)
        voltage = random.randint(20, 29)
        plant_status = "High Stress"
    elif state == 'bioelectrical':
        voltage = random.randint(8, 18)
        plant_status = "High Stress"
        
    payload = {
        "timestamp": timestamp,
        "temperature_humidity": {
            "temperature_c": temp,
            "humidity_percent": humidity
        },
        "npk": {
            "nitrogen_mg_kg": n,
            "phosphorus_mg_kg": p,
            "potassium_mg_kg": k
        },
        "dissolved_oxygen": {
            "do_mg_l": do_mg
        },
        "bioelectrical": {
            "voltage_mv": voltage,
            "plant_status": plant_status
        }
    }
    
    return payload

def main():
    print_menu()
    
    # Start keyboard listener in background thread
    listener = threading.Thread(target=read_keyboard, daemon=True)
    listener.start()
    
    while True:
        with state_lock:
            state = current_state
            
        payload = generate_telemetry(state)
        
        try:
            req = urllib.request.Request(
                API_URL, 
                data=json.dumps(payload).encode('utf-8'),
                headers={'Content-Type': 'application/json'}
            )
            with urllib.request.urlopen(req) as response:
                res_body = response.read().decode('utf-8')
                res_json = json.loads(res_body)
                
            print(
                f"[{datetime.now().strftime('%H:%M:%S')}] Sent Telemetry | State: {state.upper()} | "
                f"Temp: {payload['temperature_humidity']['temperature_c']}C | "
                f"DO: {payload['dissolved_oxygen']['do_mg_l']}mg/L | "
                f"Voltage: {payload['bioelectrical']['voltage_mv']}mV | "
                f"Status: {payload['bioelectrical']['plant_status']} | "
                f"Severity: {res_json['analysis']['severity']}"
            )
        except urllib.error.URLError as e:
            print(f"[{datetime.now().strftime('%H:%M:%S')}] Connection Error: {e.reason}")
        except Exception as e:
            print(f"[{datetime.now().strftime('%H:%M:%S')}] Error: {str(e)}")
            
        time.sleep(5)

if __name__ == "__main__":
    main()

File: simulator.js
const http = require('http');
const https = require('https');
const readline = require('readline');

const API_URL = process.argv.includes('--cloud')
  ? 'https://smart-agri-platform-1.onrender.com/api/sensor-data'
  : 'http://localhost:10000/api/sensor-data';

// Simulation state: 'normal', 'heat', 'humidity', 'hypoxia', 'bioelectrical'
let currentState = 'normal';

function printMenu() {
  console.clear();
  console.log("============================================================");
  console.log("   Agri AI IoT Hardware Simulator (Node.js Edition)  ");
  console.log("============================================================");
  console.log("Interactive Stress Controls:");
  console.log("  [n] - Reset to NORMAL state");
  console.log("  [t] - Inject HEAT stress (High Temperature)");
  console.log("  [h] - Inject DRY stress (Low Humidity)");
  console.log("  [o] - Inject HYPOXIA stress (Low Dissolved Oxygen)");
  console.log("  [s] - Inject BIOELECTRICAL stress (High Stress status)");
  console.log("  [q] / [Ctrl+C] - Quit Simulator");
  console.log("------------------------------------------------------------");
  console.log(`Current Simulation State: ${currentState.toUpperCase()}`);
  console.log("============================================================");
  console.log("Press a key at any time to inject/change stress conditions...");
}

// Enable keypress reading in TTY
readline.emitKeypressEvents(process.stdin);
if (process.stdin.isTTY) {
  process.stdin.setRawMode(true);
}

process.stdin.on('keypress', (str, key) => {
  if (key && (key.name === 'q' || (key.ctrl && key.name === 'c'))) {
    console.log('\n[SIMULATOR] Exiting...');
    process.exit();
  }

  const char = str ? str.toLowerCase() : '';
  if (['n', 't', 'h', 'o', 's'].includes(char)) {
    if (char === 'n') currentState = 'normal';
    if (char === 't') currentState = 'heat';
    if (char === 'h') currentState = 'humidity';
    if (char === 'o') currentState = 'hypoxia';
    if (char === 's') currentState = 'bioelectrical';
    
    printMenu();
    console.log(`\n[SIMULATOR] State changed to: ${currentState.toUpperCase()}`);
  }
});

function generateTelemetry(state) {
  const timestamp = new Date().toISOString();
  
  // Base normal values
  let temp = +(22 + Math.random() * 4).toFixed(1);
  let humidity = +(60 + Math.random() * 15).toFixed(1);
  let n = Math.floor(110 + Math.random() * 20);
  let p = Math.floor(25 + Math.random() * 10);
  let k = Math.floor(85 + Math.random() * 20);
  let do_mg = +(6.5 + Math.random()).toFixed(2);
  let voltage = Math.floor(45 + Math.random() * 10);
  let plant_status = "Healthy";
  
  if (state === 'heat') {
    temp = +(32 + Math.random() * 4).toFixed(1);
    voltage = Math.floor(30 + Math.random() * 12);
    plant_status = "Moderate Stress";
  } else if (state === 'humidity') {
    humidity = +(40 + Math.random() * 8).toFixed(1);
    voltage = Math.floor(35 + Math.random() * 10);
    plant_status = "Moderate Stress";
  } else if (state === 'hypoxia') {
    do_mg = +(3.5 + Math.random()).toFixed(2);
    voltage = Math.floor(20 + Math.random() * 9);
    plant_status = "High Stress";
  } else if (state === 'bioelectrical') {
    voltage = Math.floor(8 + Math.random() * 10);
    plant_status = "High Stress";
  }
  
  return {
    timestamp,
    temperature_humidity: {
      temperature_c: temp,
      humidity_percent: humidity
    },
    npk: {
      nitrogen_mg_kg: n,
      phosphorus_mg_kg: p,
      potassium_mg_kg: k
    },
    dissolved_oxygen: {
      do_mg_l: do_mg
    },
    bioelectrical: {
      voltage_mv: voltage,
      plant_status
    }
  };
}

function sendTelemetry() {
  const payload = generateTelemetry(currentState);
  const dataString = JSON.stringify(payload);
  
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(dataString)
    }
  };

  const client = API_URL.startsWith('https') ? https : http;
  const req = client.request(API_URL, options, (res) => {
    let responseBody = '';
    res.on('data', (chunk) => { responseBody += chunk; });
    res.on('end', () => {
      if (res.statusCode === 200) {
        const parsed = JSON.parse(responseBody);
        console.log(`[${new Date().toLocaleTimeString()}] Sent Telemetry | State: ${currentState.toUpperCase()} | Temp: ${payload.temperature_humidity.temperature_c}°C | DO: ${payload.dissolved_oxygen.do_mg_l}mg/L | Status: ${payload.bioelectrical.plant_status} | Severity: ${parsed.analysis.severity}`);
      } else {
        console.error(`[ERROR] Server responded with code ${res.statusCode}: ${responseBody}`);
      }
    });
  });

  req.on('error', (err) => {
    console.error(`[ERROR] Connection failed: ${err.message}`);
  });

  req.write(dataString);
  req.end();
}

// Start simulator loop
printMenu();
setInterval(sendTelemetry, 5000);
